zkevm-circuits/src/blake2b/circuit/tests/integration.rs

```diff
@@ -0,0 +1,73 @@
+     "ba80a53f981c4d0d6a2797b69f12f6e94c212f14685ac4b74b12bb6fdbffa2d17d87c5392aab792dc252d5de4533cc9518d38aa8dbf1925ab92386edd4009923"],
+    ["0000000c48c9bdf267e6096a3ba7ca8485ae67bb2bf894fe72f36e3cf1361d5f3af54fa5d182e6ad7f520e511f6c3e2b8c68059b6bbd41fbabd9831f79217e1319cde05b61626300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000000", 
+     "75ab69d3190a562c51aef8d88f1c2775876944407270c42c9844252c26d2875298743e7f6d5ea2f2d3e8d226039cd31b4e426ac4f2d3d666a610c2116fde4735"],
+    ["0000000148c9bdf267e6096a3ba7ca8485ae67bb2bf894fe72f36e3cf1361d5f3af54fa5d182e6ad7f520e511f6c3e2b8c68059b6bbd41fbabd9831f79217e1319cde05b61626300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000001", 
+     "b63a380cb2897d521994a85234ee2c181b5f844d2c624c002677e9703449d2fba551b3a8333bcdf5f2f7e08993d53923de3d64fcc68c034e717b9293fed7a421"]
+]; 
+
+// Creates the description of a BLAKE2b compression function input, which corresponds to the    
+// specified EIP-152 test vector, in accordance with the format used in the tested library
+pub fn hex_to_input(v: &str) -> CompressionInput { 
+    let mut buffer = [0u8; 213];
+    decode_to_slice(v, &mut buffer).expect("Hex input must be correct!");
+    assert!(buffer[212] <= 1, "Incorrect final block indicator flag!");
+
+    CompressionInput {
+        r: u32::from_be_bytes(buffer[0..4].try_into().unwrap()),
+        h: array::from_fn(|i| u64::from_le_bytes(buffer[4 + 8 * i..][..8].try_into().unwrap())),
+        m: array::from_fn(|i| u64::from_le_bytes(buffer[68 + 8 * i..][..8].try_into().unwrap())),
+        t: u128::from_le_bytes(buffer[196..212].try_into().unwrap()),
+        f: buffer[212] == 1
+    }
+}
+
+// Tests the circuit instance, which has height 2^17 and processes the BLAKE2b 
+// compression function inputs corresponding to the EIP-152 test vectors 4-7
+#[test]
+fn circuit_check() {
+    let k = 17;
+    let vectors: Vec<CompressionInput> = EIP152_VECTORS.iter().map(|v| hex_to_input(v[0])).collect();
+    let circuit = CompressionCircuit::<Fr,128>::new(k, &vectors);
+    let prover = MockProver::run(k, &circuit, vec![]).unwrap();
+    prover.assert_satisfied();
+}
```

Comment:
I'm missing the part of this test that we assert the output equals the expected test vector output. Did I overlook something?

Reply:
Here is a suggestion of a possible solution.

Given the `blake2b` circuit takes two advice inputs/witnesses (besides, ofc, the internal ones): the input of the hash, and its output.

First we fetch the input and output from the advice
$$(input, output) \gets \mathbb{A}$$

Then we compute the blake hash into result as the circuit synthesize
$$result \gets \mathbb{C}$$

Finally, we assert the computed result to be the expected witness
$$result \equiv output$$

This approach forces the prover to compute the expected output himself, and allows the users of the circuit to pass this `output` advice column to different circuits. With that, we can easily pass the expected input/output from the test vectors

Reply:
Actually, we do not need this, because in the current architecture the supercircuit will compute input/output pair (outside the circuit) during the interaction and hash them by means of random linear combination hash. The input/output pairs and their rlc hashes are also computed by the current circuit (inside it). The rlc hash computed by the supercircuit is going to be constrained to be into the lookup table containing all rlc's computed by the current circuit. Thus, if some i/o pair produced by this circuit is incorrect, then lookup constraint will not be satisfied.

Reply:
Understood; however, it is very expensive to test the super circuit. In case a bug is introduced here, it is preferable to test the blake circuit individually rather than depending on the full integrated test.

This is ofc not a blocker, so I'll open an issue to add a unit test in the future

Reply:
https://github.com/taikoxyz/zkevm-circuits/issues/127

Reply:
We do no need to test it in the supercircuit. The current test are sufficient. If MockProver doesn't panic during performing the test, this means that the computations performed during the circuit synthesis produce the expected output and all witness data satisfy the constraints. The correctness of the algorithm computing the expected compression function (outside the circuit) is checked by its own test.